src/core/inductive.ml

```diff
@@ -0,0 +1,244 @@
+(** Generating of induction principles. *)
+
+open Timed
+open Pos
+open Console
+open Terms
+open Print
+open Syntax
+
+(** Logging function for generating of inductive principle. *)
+let log_ind = new_logger 'g' "indu" "generating induction principle"
+let log_ind = log_ind.logger
+
+(** Builtin configuration for induction. *)
+type config =
+  { symb_Prop : sym (** : TYPE. Type of propositions. *)
+  ; symb_prf  : sym (** : Prop → TYPE.
+                        Interpretation of propositions as types. *) }
+
+(** [get_config ss pos] build the configuration using [ss]. *)
+let get_config : Sig_state.t -> Pos.popt -> config = fun ss pos ->
+  let builtin = Builtin.get ss pos in
+  { symb_Prop = builtin "Prop"
+  ; symb_prf  = builtin "P" }
+
+(** [check_typ_ind rec_typ] checks if the type of the term [rec_typ] is the
+    constant TYPE. *)
+let check_type_ind : popt -> term -> unit = fun pos rec_typ ->
+  match Typing.infer [] rec_typ with
+  | Some t ->
+      begin
+      match unfold t with
+      | Type   -> ()
+      | _ -> fatal pos "The type of the generated inductive principle of [%a]
+                        isn't the constant TYPE. Please, raise an issue."
+               pp_term rec_typ
+      end
+  | None   ->
+      fatal pos "The type of the generated inductive principle of
+                 [%a] isn't typable. Please, raise an issue."
+        pp_term rec_typ
+
+(** [gen_ind_typ_codom pos ind_sym codom s] assumes that the type of [ind_sym]
+    is of the form [Π(i1:a1),...,Π(in:an), TYPE]. It then generates a [tbox]
+    similar to this type except that [TYPE] is replaced by
+    [codom [i1;...;in]]. The string [s] is the prefix of variables' name. *)
+let gen_ind_typ_codom : popt -> sym -> (tbox list -> tbox) -> string -> tbox =
+  fun pos ind_sym codom s ->
+  let rec aux : tvar list -> term -> tbox = fun xs a ->
+    match Basics.get_args a with
+    | (Type, _) -> codom (List.rev_map _Vari xs)
+    | (Prod(a,b), _) ->
+        let (x,b) = Basics.unbind_name b s in
+        _Prod (lift a) (Bindlib.bind_var x (aux (x::xs) b))
+    | _ -> fatal pos "The type of %a is not supported" pp_symbol ind_sym
+  in aux [] !(ind_sym.sym_type)
+
+(** [fold_cons_typ pos codom inj_var build_rec_hyp domrec dom ind_sym cons_sym
+    f_rec f_not_rec init s] generates some value iteratively by going through
+    the structure of [cons_sym.sym_type]. The argument [pos] is the position
```

Comment:
You should say that it is equal to `aux [] init !(cons_sym.sym_type)` where `aux : 'b list -> 'a -> term -> 'c`, and explain aux and next.

Reply:
What's the reference to "it"?

Reply:
fold_cons_typ